pubsub.go

```diff
@@ -42,71 +42,95 @@ type PubsubValueStore struct {
+	mx     sync.Mutex
+	topics map[string]*topicInfo
 
 	watchLk  sync.Mutex
 	watching map[string]*watchGroup
 
 	Validator record.Validator
 }
 
+type topicInfo struct {
+	topic *pubsub.Topic
+	evts  *pubsub.TopicEventHandler
+	sub   *pubsub.Subscription
+
+	cancel   context.CancelFunc
+	finished chan struct{}
+
+	dbWriteMx sync.Mutex
+}
+
 // KeyToTopic converts a binary record key to a pubsub topic key.
 func KeyToTopic(key string) string {
 	// Record-store keys are arbitrary binary. However, pubsub requires UTF-8 string topic IDs.
 	// Encodes to "/record/base64url(key)"
 	return "/record/" + base64.RawURLEncoding.EncodeToString([]byte(key))
 }
 
-// NewPubsubPublisher constructs a new Publisher that publishes IPNS records through pubsub.
-// The constructor interface is complicated by the need to bootstrap the pubsub topic.
-// This could be greatly simplified if the pubsub implementation handled bootstrap itself
-func NewPubsubValueStore(ctx context.Context, host host.Host, cr routing.ContentRouting, ps *pubsub.PubSub, validator record.Validator) *PubsubValueStore {
+type PubsubValueStoreOpt func(*PubsubValueStore) error
+
+// NewPubsubValueStore constructs a new ValueStore that gets and receives records through pubsub.
+func NewPubsubValueStore(ctx context.Context, host host.Host, ps *pubsub.PubSub, validator record.Validator, opts ...PubsubValueStoreOpt) (*PubsubValueStore, error) {
 	psValueStore := &PubsubValueStore{
 		ctx: ctx,
 
-		cr: cr, // needed for pubsub bootstrap
-
 		ds:                      dssync.MutexWrap(ds.NewMapDatastore()),
 		ps:                      ps,
 		host:                    host,
 		rebroadcastInitialDelay: 100 * time.Millisecond,
 		rebroadcastInterval:     time.Minute * 10,
 
-		subs:     make(map[string]*pubsub.Subscription),
-		cancels:  make(map[string]context.CancelFunc),
+		topics:   make(map[string]*topicInfo),
 		watching: make(map[string]*watchGroup),
 
 		Validator: validator,
 	}
 
+	for _, opt := range opts {
+		err := opt(psValueStore)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	psValueStore.fetch = newFetchProtocol(ctx, host, psValueStore.getLocal)
 
 	go psValueStore.rebroadcast(ctx)
 
-	return psValueStore
+	return psValueStore, nil
 }
 
-// Publish publishes an IPNS record through pubsub with default TTL
+// PutValue publishes a record through pubsub
 func (p *PubsubValueStore) PutValue(ctx context.Context, key string, value []byte, opts ...routing.Option) error {
-	// Record-store keys are arbitrary binary. However, pubsub requires UTF-8 string topic IDs.
-	// Encode to "/record/base64url(key)"
-	topic := KeyToTopic(key)
-
 	if err := p.Subscribe(key); err != nil {
 		return err
 	}
 
 	log.Debugf("PubsubPublish: publish value for key", key)
 
+	p.mx.Lock()
+	ti, ok := p.topics[key]
+	if !ok {
+		p.mx.Unlock()
+		return errors.New("could not find topic handle")
+	}
+	p.mx.Unlock()
+
+	ti.dbWriteMx.Lock()
+	defer ti.dbWriteMx.Unlock()
+	ok, err := p.putLocal(ti, key, value)
```

Comment:
Previously we were receiving updates by waiting for `pubsub.Publish(topic,value)` to return us a value via our own `Subscription`. However, pubsub can drop messages so this isn't great, we now actually do a local put before the pubsub Put.

This does mean that we'll receive most of our messages twice (and therefore deserialize, validate and compare them to the most recent state twice).

Reply:
We shouldn't be dropping locally published messages. But yeah, it makes sense to put them locally anyways just in case.